java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.compare;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.arrow.memory.util.ByteFunctionHelpers;
+import org.apache.arrow.vector.BaseFixedWidthVector;
+import org.apache.arrow.vector.BaseVariableWidthVector;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.ValueVector;
+import org.apache.arrow.vector.ZeroVector;
+import org.apache.arrow.vector.complex.BaseRepeatedValueVector;
+import org.apache.arrow.vector.complex.FixedSizeListVector;
+import org.apache.arrow.vector.complex.ListVector;
+import org.apache.arrow.vector.complex.NonNullableStructVector;
+import org.apache.arrow.vector.complex.UnionVector;
+
+/**
+ * Visitor to compare a range of values for vectors.
+ */
+public class RangeEqualsVisitor {
+
+  protected final ValueVector right;
+  protected final int leftStart;
+  protected final int rightStart;
+  protected final int length;
+
+  /**
+   * Constructs a new instance.
+   */
+  public RangeEqualsVisitor(ValueVector right, int leftStart, int rightStart, int length) {
+    this.leftStart = leftStart;
+    this.rightStart = rightStart;
+    this.right = right;
+    this.length = length;
+  }
+
+  public boolean visit(BaseFixedWidthVector left) {
+    return compareBaseFixedWidthVectors(left);
+  }
+
+  public boolean visit(BaseVariableWidthVector left) {
+    return compareBaseVariableWidthVectors(left);
+  }
+
+  public boolean visit(ListVector left) {
+    return compareListVectors(left);
+  }
+
+  public boolean visit(FixedSizeListVector left) {
+    return compareFixedSizeListVectors(left);
+  }
+
+  public boolean visit(NonNullableStructVector left) {
+    return compareStructVectors(left);
+  }
+
+  public boolean visit(UnionVector left) {
+    return compareUnionVectors(left);
+  }
+
+  public boolean visit(ZeroVector left) {
+    return false;
+  }
+
+  protected boolean compareUnionVectors(UnionVector left) {
+    if (!left.getField().getType().equals(right.getField().getType())) {
+      return false;
+    }
+
+    UnionVector rightVector = (UnionVector) right;
```

Comment:
Is this a safe cast? I guess you are comparing Field type before, so they _should_ be the same vector class

Reply:
Right.